# Imports and setup

In [1]:
import pickle
import numpy as np
import pandas as pd

# Reading the dictionary file

In [ ]:
# Adjust the path accordingly to where the file is located on your system

with open('data/raw/complete_NCBI_dict', 'rb') as file:
    NCBI_dict = pickle.load(file)

In [3]:
print("The dictionary contains data on {} participants (1 sample per participant)".format(len(NCBI_dict.keys())))

The dictionary contains data on 28017 participants (1 sample per participant)


In [4]:
# The dictionary is setup with participant / sample IDs as keys. You can access these IDs as follows:

NCBI_dict.keys()

dict_keys(['ERR1073890', 'ERR1073895', 'ERR1073904', 'ERR1073906', 'ERR1073910', 'ERR1073916', 'ERR1073918', 'ERR1073921', 'ERR1073928', 'ERR1073929', 'ERR1073937', 'ERR1073940', 'ERR1073946', 'ERR1073952', 'ERR1073970', 'ERR1073977', 'ERR1073979', 'ERR1073993', 'ERR1074018', 'ERR1079918', 'ERR1079921', 'ERR1080005', 'ERR1080008', 'ERR1080025', 'ERR1080029', 'ERR1080035', 'ERR1080044', 'ERR1080049', 'ERR1080051', 'ERR1080062', 'ERR1080073', 'ERR1080078', 'ERR1080087', 'ERR1080105', 'ERR1080121', 'ERR1316228', 'ERR1316229', 'ERR1316230', 'ERR1316245', 'ERR1316252', 'ERR1316253', 'ERR1316255', 'ERR1316258', 'ERR1316264', 'ERR1316265', 'ERR1316266', 'ERR1316269', 'ERR1316272', 'ERR1316276', 'ERR1316280', 'ERR1316284', 'ERR1316291', 'ERR1316293', 'ERR1316302', 'ERR1846202', 'ERR1846205', 'ERR1846210', 'ERR1846217', 'ERR1852407', 'ERR1852424', 'ERR1852430', 'ERR1852436', 'ERR1852439', 'ERR1852440', 'ERR1852444', 'ERR1852454', 'ERR1852464', 'ERR1854662', 'ERR1854663', 'ERR1854665', 'ERR18546

In [5]:
# Each ID contain various sub dictionaries / lists with different data categories. You can access these categories as follows:

NCBI_dict['ERR1073890'].keys()

dict_keys(['Biosample', 'Attributes', 'Description', 'Accession_ID', 'Phylum', 'Taxa'])

# Structure of the dataset

```
NCBI_dict
│
├── Sample_ID_1
│   ├── Biosample              (reference ID linking to NCBI BioSample database)
│   ├── Attributes             (metadata)
│   ├── Description            (sample description)
│   ├── Accession_ID           (sequence accession number from NCBI - SAR)
│   ├── Phylum                 (aggregated taxonomic counts at the phylum level)
│   └── Taxa                   (taxonomic data at species level)
│
├── Sample_ID_2
│   ├── Biosample
│   ├── Attributes
│   ├── Description
│   ├── Accession_ID
│   ├── Phylum
│   └── Taxa
│
└── Sample_ID_N
    ├── Biosample
    ├── Attributes
    ├── Description
    ├── Accession_ID
    ├── Phylum
    └── Taxa
```

Biosample: Links to metadata about the sample itself (origin, collection date, body site, health status, etc.)

Accession_ID: Links to the actual DNA/RNA sequences - SRA: Sequence Read Archive, which is NCBI's repository for raw sequencing data

In [6]:
NCBI_dict['ERR1073890']['Biosample']

'SAMEA3608835'

In [7]:
NCBI_dict['ERR1073910']['Description']

'American Gut Project Nares Sample'

In [8]:
NCBI_dict['ERR1073910']['Accession_ID']

'SAMEA360885'

# To access meta data for a sample ID

In [9]:
NCBI_dict['ERR1073910']['Attributes']

{'ENA first public': '2015-10-15',
 'ENA last update': '2017-08-10',
 'ENA-CHECKLIST': 'ERC000011',
 'External Id': 'SAMEA3608855',
 'INSDC center alias': 'UCSDMI',
 'INSDC center name': 'University of California San Diego Microbiome Initiative',
 'INSDC first public': '2015-10-15T17:01:16Z',
 'INSDC last update': '2017-08-10T16:56:16Z',
 'INSDC status': 'public',
 'Submitter Id': 'qiita_sid_10317:10317.000005785',
 'acid_reflux': 'Not provided',
 'acne medication': 'false',
 'acne_medication_otc': 'false',
 'add_adhd': 'Not provided',
 'age cat': '40s',
 'age years': '47.0',
 'age_corrected': '47.0',
 'alcohol consumption': 'false',
 'alcohol_frequency': 'Never',
 'alcohol_types': 'Not provided',
 'alcohol_types_beercider': 'false',
 'alcohol_types_red_wine': 'false',
 'alcohol_types_sour_beers': 'false',
 'alcohol_types_spiritshard_alcohol': 'false',
 'alcohol_types_unspecified': 'true',
 'alcohol_types_white_wine': 'false',
 'allergic to i have no food allergies that i know of': 'fa

# Gut-related samples

## Cheking unique values of body site attributes
The attributes were selected based on descriptions from table S2 from supplementary materials of the American Gut Project paper. These attributes define the region of the body the sample was sampled from.

Resource: McDonald, D., Hyde, E., Debelius, J. W., Morton, J. T., Gonzalez, A., Ackermann, G., Aksenov, A. A., Behsaz, B., Brennan, C., Chen, Y., DeRight Goldasich, L., Dorrestein, P. C., Dunn, R. R., Fahimipour, A. K., Gaffney, J., Gilbert, J. A., Gogul, G., Green, J. L., Hugenholtz, P., … Gunderson, B. (2018). American Gut: An Open Platform for Citizen Science Microbiome Research. mSystems, 3(3), e00031-18. https://doi.org/10.1128/mSystems.00031-18

In [32]:
# Collect and display unique values for each attribute
print("\n\n[ATTRIBUTE VALUES] Unique values in dataset for each body site attribute")
print("-" * 120)

type_attributes = ['description', 'environmental package', 'environmental medium', 'body habitat', 'body product', 'sample type']

unique_values = {}
for attr_name in type_attributes:
    unique_vals = set()
    for sample_id in NCBI_dict.keys():
        if attr_name in NCBI_dict[sample_id]['Attributes']:
            value = NCBI_dict[sample_id]['Attributes'][attr_name]
            # Handle both string and list values
            if isinstance(value, list):
                unique_vals.update(value)
            else:
                unique_vals.add(value)
    unique_values[attr_name] = sorted(unique_vals)
    
    print(f"\n{attr_name}:")
    if unique_vals:
        for val in sorted(unique_vals):
            print(f"  • {val}")
    else:
        print(f"  (no values found)")



[ATTRIBUTE VALUES] Unique values in dataset for each body site attribute
------------------------------------------------------------------------------------------------------------------------

description:
  • American Gut Project Ear Wax Sample
  • American Gut Project Forehead Sample
  • American Gut Project Hair Sample
  • American Gut Project Left Hand Sample
  • American Gut Project Mouth Sample
  • American Gut Project Nares Sample
  • American Gut Project Nasal Mucus Sample
  • American Gut Project Right Hand Sample
  • American Gut Project Stool Sample
  • American Gut Project Vaginal Mucus Sample
  • American Gut control

environmental package:
  • Unspecified
  • air
  • host-associated
  • human-associated
  • human-gut
  • human-oral
  • human-skin
  • human-vaginal
  • misc environment

environmental medium:
  • LabControl test
  • Unspecified
  • blood
  • ear wax
  • feces
  • food product
  • mucus
  • saliva
  • sebum
  • sterile water
  • tears

body habitat:
  • 

## Presence of body site attributes in the dataset

In [34]:
print("=" * 120)
print("GUT MICROBIOME SAMPLE ANALYSIS: Body Site Attribute Coverage & Value Consistency")
print("=" * 120)

# ============================================================================
# ANALYSIS 1: ATTRIBUTE COVERAGE - How many samples have each attribute?
# ============================================================================
print("\n[ANALYSIS 1] BODY SITE ATTRIBUTE COVERAGE")
print("-" * 120)

attribute_coverage = {}
for attr_name in type_attributes:
    count = 0
    for sample_id in NCBI_dict.keys():
        if attr_name in NCBI_dict[sample_id]['Attributes']:
            count += 1
    attribute_coverage[attr_name] = {
        'count': count,
        'percentage': (count / len(NCBI_dict)) * 100
    }

print(f"\n{'Attribute':<35} {'# Samples':<15} {'% Coverage':<15}")
print("-" * 120)
for attr_name, data in attribute_coverage.items():
    print(f"{attr_name:<35} {data['count']:<15} {data['percentage']:<14.1f}%")

# ============================================================================
# ANALYSIS 2: ATTRIBUTE COMBINATION - How many attributes does each sample have?
# ============================================================================
print("\n\n[ANALYSIS 2] BODY SITE ATTRIBUTE COMBINATION DISTRIBUTION")
print("-" * 120)

# Count how many gut-related attributes each sample has
attribute_counts_per_sample = {}
for sample_id in NCBI_dict.keys():
    attributes = NCBI_dict[sample_id]['Attributes']
    count = sum(1 for attr_name in type_attributes if attr_name in attributes)
    
    if count not in attribute_counts_per_sample:
        attribute_counts_per_sample[count] = 0
    attribute_counts_per_sample[count] += 1

print(f"\n{'# Attributes Present':<25} {'# Samples':<15} {'% of Total':<15}")
print("-" * 120)
for num_attrs in sorted(attribute_counts_per_sample.keys()):
    count = attribute_counts_per_sample[num_attrs]
    percentage = (count / len(NCBI_dict)) * 100
    print(f"{num_attrs:<25} {count:<15} {percentage:<14.1f}%")

samples_with_all_6 = attribute_counts_per_sample.get(6, 0)
print(f"\n✓ Samples with ALL 6 attributes: {samples_with_all_6} ({samples_with_all_6/len(NCBI_dict)*100:.1f}%)")

GUT MICROBIOME SAMPLE ANALYSIS: Body Site Attribute Coverage & Value Consistency

[ANALYSIS 1] BODY SITE ATTRIBUTE COVERAGE
------------------------------------------------------------------------------------------------------------------------

Attribute                           # Samples       % Coverage     
------------------------------------------------------------------------------------------------------------------------
description                         1643            5.9           %
environmental package               21689           77.4          %
environmental medium                22434           80.1          %
body habitat                        19679           70.2          %
body product                        19623           70.0          %
sample type                         23030           82.2          %


[ANALYSIS 2] BODY SITE ATTRIBUTE COMBINATION DISTRIBUTION
-------------------------------------------------------------------------------------------------

We can see that not all samples include all attributes. Some samples don't even have any of the 6 attributes.

For each body site attribute above these are the values that relate to gut microbiome samples:
```
ATTRIBUE                        VALUE

description                     American Gut Project Stool Sample

environmental package	        human-gut

environmental medium	        feces

body habitat	                UBERON:feces

body product	                UBERON:feces

sample type                     stool OR feces
```

## Consistency in expected values for the 6 attributes

In [12]:
# ============================================================================
# CONSISTENCY ANALYSIS: Identify inconsistent samples
# ============================================================================
# For each sample, check ONLY the attributes that are present and SPECIFIED.
# "Not provided" values are IGNORED (not counted as mismatches).
# Uses MAJORITY VOTE: if >50% of specified values match expected, sample is consistent.

print("\n\n" + "=" * 120)
print("CONSISTENCY ANALYSIS: Majority Vote with 'Not Provided' Ignored")
print("=" * 120)

# Define gut attributes with acceptable values (case-insensitive matching)
gut_attributes = {
    'description': 'American Gut Project Stool Sample',
    'environmental package': 'human-gut',
    'environmental medium': 'feces',
    'body habitat': 'UBERON:feces',
    'body product': 'UBERON:feces',
    'sample type': ['Stool', 'feces']  # Accept both Stool and feces (case-insensitive)
}

def value_matches(actual_value, expected_value):
    """
    Check if actual value matches expected value.
    Handles both single values and lists.
    Case-insensitive matching.
    """
    if actual_value.lower() == 'not provided':
        return None  # None means "not specified, ignore"
    
    if isinstance(expected_value, list):
        # For lists, check if actual matches any item (case-insensitive)
        return any(actual_value.lower() == exp.lower() for exp in expected_value)
    else:
        # For single values, case-insensitive comparison
        return actual_value.lower() == expected_value.lower()

# Analyze each sample
perfect_100_samples = []     # 100% match
partial_50_to_100_samples = []  # >50% but <100% match
partial_match_samples = []   # Some match, some don't (but not majority)
complete_mismatch_samples = []  # None match or no specified values to evaluate

sample_details = {}  # Store details of all non-perfect samples

for sample_id in NCBI_dict.keys():
    attributes = NCBI_dict[sample_id]['Attributes']
    
    # Find which gut attributes are present in this sample
    present_attrs = {k: v for k, v in attributes.items() if k in gut_attributes}
    
    if not present_attrs:
        # Skip samples with none of the gut attributes
        continue
    
    # Check which present attributes match expected values
    # Ignore "Not provided" values
    matches = 0
    specified_count = 0  # Count of non-"not provided" attributes
    attr_results = {}
    
    for attr_name, actual_value in present_attrs.items():
        expected_value = gut_attributes[attr_name]
        match_result = value_matches(actual_value, expected_value)
        
        # Format expected value for display
        if isinstance(expected_value, list):
            expected_display = " or ".join([f"'{v}'" for v in expected_value])
        else:
            expected_display = expected_value
        
        attr_results[attr_name] = {
            'expected': expected_display,
            'actual': actual_value,
            'match': match_result  # True, False, or None (not specified)
        }
        
        if match_result is None:
            # "Not provided" - ignore it
            pass
        elif match_result:
            matches += 1
            specified_count += 1
        else:
            specified_count += 1
    
    # Calculate match percentage
    match_percentage = (matches / specified_count * 100) if specified_count > 0 else 0
    
    # Categorize sample using majority vote
    if specified_count == 0:
        # All attributes are "Not provided" - can't evaluate
        complete_mismatch_samples.append(sample_id)
        sample_details[sample_id] = attr_results
    elif match_percentage == 100:
        # Perfect match (100%) - PERFECT
        perfect_100_samples.append(sample_id)
    elif matches >= specified_count / 2:
        # Majority match (>50% but <100%) - CONSISTENT
        partial_50_to_100_samples.append(sample_id)
    elif matches == 0:
        # No matches at all - COMPLETE MISMATCH
        complete_mismatch_samples.append(sample_id)
        sample_details[sample_id] = attr_results
    else:
        # Some match, but not majority - PARTIAL
        partial_match_samples.append(sample_id)
        sample_details[sample_id] = attr_results

# Print statistics
print(f"\n[SUMMARY] Sample Consistency Breakdown (Case-Insensitive, Majority Vote)")
print("-" * 120)
total_samples = len(perfect_100_samples) + len(partial_50_to_100_samples) + len(partial_match_samples) + len(complete_mismatch_samples)
print(f"\nTotal samples with any gut attributes: {total_samples:,}")
print(f"\n{'Category':<50} {'# Samples':<20} {'% of Total':<20}")
print("-" * 120)
print(f"{'Perfect Match (100%)':<50} {len(perfect_100_samples):<20} {len(perfect_100_samples)/total_samples*100:<19.1f}%")
print(f"{'Consistent (>50% to <100%)':<50} {len(partial_50_to_100_samples):<20} {len(partial_50_to_100_samples)/total_samples*100:<19.1f}%")
print(f"{'TOTAL ≥50% MATCH (GUT SAMPLES)':<50} {len(perfect_100_samples) + len(partial_50_to_100_samples):<20} {(len(perfect_100_samples) + len(partial_50_to_100_samples))/total_samples*100:<19.1f}%")
print(f"{'Partial Match (<50%)':<50} {len(partial_match_samples):<20} {len(partial_match_samples)/total_samples*100:<19.1f}%")
print(f"{'Complete Mismatch (0% match)':<50} {len(complete_mismatch_samples):<20} {len(complete_mismatch_samples)/total_samples*100:<19.1f}%")

# ============================================================================
# DETAILED LOOK AT PARTIAL MATCH SAMPLES
# ============================================================================
print(f"\n\n[PARTIAL MATCHES] Samples with Some Matches (<50%)")
print("-" * 120)

print(f"\nTotal partial match samples: {len(partial_match_samples)}")
print(f"\nShowing first 30 samples:\n")

for idx, sample_id in enumerate(partial_match_samples[:30]):
    attr_results = sample_details[sample_id]
    
    # Count matches/mismatches/unspecified for this sample
    matches_count = sum(1 for r in attr_results.values() if r['match'] is True)
    mismatches_count = sum(1 for r in attr_results.values() if r['match'] is False)
    unspecified_count = sum(1 for r in attr_results.values() if r['match'] is None)
    
    print(f"\n[Sample {idx+1}] {sample_id}")
    print(f"  Status: {matches_count} match, {mismatches_count} mismatch, {unspecified_count} not provided")
    print(f"  ─" * 60)
    
    for attr_name in sorted(attr_results.keys()):
        result = attr_results[attr_name]
        if result['match'] is None:
            status = "⊘ NOT PROVIDED"
        elif result['match']:
            status = "✓ MATCH"
        else:
            status = "✗ MISMATCH"
        
        actual_display = result['actual'][:50] + "..." if len(result['actual']) > 50 else result['actual']
        
        print(f"  {attr_name:<35} {status}")
        print(f"    Expected: {result['expected']}")
        print(f"    Actual:   {actual_display}")

if len(partial_match_samples) > 30:
    print(f"\n... and {len(partial_match_samples) - 30} more partial match samples")

# ============================================================================
# COMPLETE MISMATCH ANALYSIS - First 20 samples
# ============================================================================
print(f"\n\n[COMPLETE MISMATCHES] Samples with No Matches")
print("-" * 120)

print(f"\nTotal complete mismatch samples: {len(complete_mismatch_samples):,}")
print(f"\nShowing first 5 samples:\n")

for idx, sample_id in enumerate(complete_mismatch_samples[:5]):
    attr_results = sample_details.get(sample_id, {})
    
    if not attr_results:
        print(f"\n[Sample {idx+1}] {sample_id}")
        print(f"  Status: No gut attributes to evaluate")
        continue
    
    # Count matches/mismatches/unspecified for this sample
    matches_count = sum(1 for r in attr_results.values() if r['match'] is True)
    mismatches_count = sum(1 for r in attr_results.values() if r['match'] is False)
    unspecified_count = sum(1 for r in attr_results.values() if r['match'] is None)
    
    print(f"\n[Sample {idx+1}] {sample_id}")
    print(f"  Status: {matches_count} match, {mismatches_count} mismatch, {unspecified_count} not provided")
    print(f"  ─" * 60)
    
    for attr_name in sorted(attr_results.keys()):
        result = attr_results[attr_name]
        if result['match'] is None:
            status = "⊘ NOT PROVIDED"
        elif result['match']:
            status = "✓ MATCH"
        else:
            status = "✗ MISMATCH"
        
        actual_display = result['actual'][:50] + "..." if len(result['actual']) > 50 else result['actual']
        
        print(f"  {attr_name:<35} {status}")
        print(f"    Expected: {result['expected']}")
        print(f"    Actual:   {actual_display}")

if len(complete_mismatch_samples) > 5:
    print(f"\n... and {len(complete_mismatch_samples) - 5} more complete mismatch samples")

# ============================================================================
# ANALYZE PATTERNS IN PARTIAL MATCHES
# ============================================================================
print(f"\n\n[PATTERNS] Common Attribute Patterns in Partial Matches")
print("-" * 120)

# For partial match samples, see which attributes typically match/mismatch
mismatch_frequency = {}
match_frequency = {}
unspecified_frequency = {}

for sample_id in partial_match_samples:
    attr_results = sample_details[sample_id]
    for attr_name, result in attr_results.items():
        if result['match'] is True:
            if attr_name not in match_frequency:
                match_frequency[attr_name] = 0
            match_frequency[attr_name] += 1
        elif result['match'] is False:
            if attr_name not in mismatch_frequency:
                mismatch_frequency[attr_name] = 0
            mismatch_frequency[attr_name] += 1
        else:  # None - not provided
            if attr_name not in unspecified_frequency:
                unspecified_frequency[attr_name] = 0
            unspecified_frequency[attr_name] += 1

print(f"\nAmong {len(partial_match_samples)} partial match samples:\n")
print(f"{'Attribute':<35} {'Matches':<15} {'Mismatches':<15} {'Not Provided':<15} {'Match %':<10}")
print("-" * 120)

for attr_name in sorted(gut_attributes.keys()):
    matches = match_frequency.get(attr_name, 0)
    mismatches = mismatch_frequency.get(attr_name, 0)
    unspecified = unspecified_frequency.get(attr_name, 0)
    total = matches + mismatches + unspecified
    
    if total > 0:
        match_pct = (matches / (matches + mismatches) * 100) if (matches + mismatches) > 0 else 0
        print(f"{attr_name:<35} {matches:<15} {mismatches:<15} {unspecified:<15} {match_pct:<9.1f}%")




CONSISTENCY ANALYSIS: Majority Vote with 'Not Provided' Ignored

[SUMMARY] Sample Consistency Breakdown (Case-Insensitive, Majority Vote)
------------------------------------------------------------------------------------------------------------------------

Total samples with any gut attributes: 23,621

Category                                           # Samples            % of Total          
------------------------------------------------------------------------------------------------------------------------
Perfect Match (100%)                               17306                73.3               %
Consistent (>50% to <100%)                         14                   0.1                %
TOTAL ≥50% MATCH (GUT SAMPLES)                     17320                73.3               %
Partial Match (<50%)                               0                    0.0                %
Complete Mismatch (0% match)                       6301                 26.7               %


[PARTIAL M

The 14 samples classified as "inconsistent" were inspected separately and showed that the mismatched attributes had values such as "not provided" or "not collected", thus the >50% rule was implemented.

Based on this analysis we can conclude that there are **17,320** samples that can be classified as gut microbiome samples and **6,301** samples that originate from other body regions. That does not add up to the total number of samples in the dataset **(28,017)**, indicating that there are **4,396** samples that don't have any of the 6 attributes present among their attributes and therefore remain unclassified in this research. 

# Filtering gut dictionary

In [14]:
# Combine perfect match and partial match (>50%) samples
gut_sample_ids = perfect_100_samples + partial_50_to_100_samples

# Create filtered dictionary containing only gut samples
gut_NCBI_dict = {sample_id: NCBI_dict[sample_id] for sample_id in gut_sample_ids}

print("=" * 100)
print("FILTERED GUT-ONLY DATASET")
print("=" * 100)
print(f"\nOriginal NCBI_dict: {len(NCBI_dict):,} samples")
print(f"Filtered gut_NCBI_dict: {len(gut_NCBI_dict):,} samples")
print(f"  ├─ Perfect match (100%): {len(perfect_100_samples):,}")
print(f"  └─ Consistent match (>50%): {len(partial_50_to_100_samples):,}")
print(f"\nReduction: {(1 - len(gut_NCBI_dict)/len(NCBI_dict))*100:.1f}% of samples removed")

FILTERED GUT-ONLY DATASET

Original NCBI_dict: 28,017 samples
Filtered gut_NCBI_dict: 17,320 samples
  ├─ Perfect match (100%): 17,306
  └─ Consistent match (>50%): 14

Reduction: 38.2% of samples removed


# The most reported attributes

## For the entire dataset

In [14]:
# Find which attributes have actual data (not "Not provided") in most samples
print("=== Attributes with Most Data Coverage ===")
print("Finding which attributes have actual (non-'Not provided') values...\n")

# Count actual data for each attribute across all samples
attribute_data_counts = {}
total_samples_with_attribute = {}

for sample_id in NCBI_dict.keys():
    attributes = NCBI_dict[sample_id]['Attributes']
    for attr_name, attr_value in attributes.items():
        # Count total occurrences of this attribute
        if attr_name not in total_samples_with_attribute:
            total_samples_with_attribute[attr_name] = 0
        total_samples_with_attribute[attr_name] += 1
        
        # Count actual data (non-"Not provided") occurrences
        if attr_value != 'Not provided':
            if attr_name not in attribute_data_counts:
                attribute_data_counts[attr_name] = 0
            attribute_data_counts[attr_name] += 1

# Calculate coverage and sort by most data
attribute_coverage = []
for attr_name in total_samples_with_attribute.keys():
    data_count = attribute_data_counts.get(attr_name, 0)
    total_count = total_samples_with_attribute[attr_name]
    coverage_percentage = (data_count / total_count) * 100
    attribute_coverage.append({
        'attribute': attr_name,
        'data_count': data_count,
        'total_count': total_count,
        'coverage_percentage': coverage_percentage
    })

# Sort by data count (descending)
attribute_coverage.sort(key=lambda x: x['data_count'], reverse=True)

print(f"{'Rank':<4} {'Attribute':<50} {'With Data':<12} {'Total':<8} {'Coverage %':<10}")
print("-" * 90)

for i, stat in enumerate(attribute_coverage[:30]):
    print(f"{i+1:<4} {stat['attribute']:<50} {stat['data_count']:<12} {stat['total_count']:<8} {stat['coverage_percentage']:<10.1f}")

# Summary statistics
print(f"\n=== Coverage Summary ===")
print(f"Total attributes: {len(attribute_coverage)}")
print(f"Attributes with >90% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 90])}")
print(f"Attributes with >50% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 50])}")
print(f"Attributes with >10% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 10])}")

=== Attributes with Most Data Coverage ===
Finding which attributes have actual (non-'Not provided') values...

Rank Attribute                                          With Data    Total    Coverage %
------------------------------------------------------------------------------------------
1    INSDC center name                                  28017        28017    100.0     
2    INSDC first public                                 28017        28017    100.0     
3    INSDC last update                                  28017        28017    100.0     
4    INSDC status                                       28017        28017    100.0     
5    sample name                                        28015        28015    100.0     
6    ENA-CHECKLIST                                      26374        26374    100.0     
7    External Id                                        26374        26374    100.0     
8    Submitter Id                                       26374        26374    100.0  

## For the gut subset dataset

In [15]:
# Find which attributes have actual data (not "Not provided") in most samples
print("=== Attributes with Most Data Coverage ===")
print("Finding which attributes have actual (non-'Not provided') values...\n")

# Count actual data for each attribute across all samples
attribute_data_counts = {}
total_samples_with_attribute = {}

for sample_id in gut_NCBI_dict.keys():
    attributes = gut_NCBI_dict[sample_id]['Attributes']
    for attr_name, attr_value in attributes.items():
        # Count total occurrences of this attribute
        if attr_name not in total_samples_with_attribute:
            total_samples_with_attribute[attr_name] = 0
        total_samples_with_attribute[attr_name] += 1
        
        # Count actual data (non-"Not provided") occurrences
        if attr_value != 'Not provided':
            if attr_name not in attribute_data_counts:
                attribute_data_counts[attr_name] = 0
            attribute_data_counts[attr_name] += 1

# Calculate coverage and sort by most data
attribute_coverage = []
for attr_name in total_samples_with_attribute.keys():
    data_count = attribute_data_counts.get(attr_name, 0)
    total_count = total_samples_with_attribute[attr_name]
    coverage_percentage = (data_count / total_count) * 100
    attribute_coverage.append({
        'attribute': attr_name,
        'data_count': data_count,
        'total_count': total_count,
        'coverage_percentage': coverage_percentage
    })

# Sort by data count (descending)
attribute_coverage.sort(key=lambda x: x['data_count'], reverse=True)

print(f"{'Rank':<4} {'Attribute':<50} {'With Data':<12} {'Total':<8} {'Coverage %':<10}")
print("-" * 90)

for i, stat in enumerate(attribute_coverage[:30]):
    print(f"{i+1:<4} {stat['attribute']:<50} {stat['data_count']:<12} {stat['total_count']:<8} {stat['coverage_percentage']:<10.1f}")

# Summary statistics
print(f"\n=== Coverage Summary ===")
print(f"Total attributes: {len(attribute_coverage)}")
print(f"Attributes with >90% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 90])}")
print(f"Attributes with >50% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 50])}")
print(f"Attributes with >10% coverage: {len([a for a in attribute_coverage if a['coverage_percentage'] > 10])}")

=== Attributes with Most Data Coverage ===
Finding which attributes have actual (non-'Not provided') values...

Rank Attribute                                          With Data    Total    Coverage %
------------------------------------------------------------------------------------------
1    INSDC center name                                  17320        17320    100.0     
2    INSDC first public                                 17320        17320    100.0     
3    INSDC last update                                  17320        17320    100.0     
4    INSDC status                                       17320        17320    100.0     
5    sample name                                        17320        17320    100.0     
6    race                                               16986        17305    98.2      
7    alcohol_frequency                                  16961        17300    98.0      
8    pool_frequency                                     16908        17303    97.7   

From now on the EDA will focus only on the gut dataset, since this is what this research is interested in.

# Count "Not provided" values

In [16]:
# Analyze "Not provided" values for all samples
print(f"Analyzing 'Not provided' values for all {len(gut_NCBI_dict)} samples...")

def count_not_provided_values(sample_id):
    """Count how many attributes have 'Not provided' values for a given sample"""
    if sample_id not in gut_NCBI_dict:
        return None, None
    
    attributes = gut_NCBI_dict[sample_id]['Attributes']
    total_attributes = len(attributes)
    not_provided_count = 0
    
    for attr_name, attr_value in attributes.items():
        if attr_value == 'Not provided':
            not_provided_count += 1
    
    return not_provided_count, total_attributes

# Store results for analysis
not_provided_results = {}
sample_stats = []

for sample_id in gut_NCBI_dict.keys():
    not_provided, total = count_not_provided_values(sample_id)
    if not_provided is not None:
        percentage = (not_provided / total) * 100
        not_provided_results[sample_id] = {
            'not_provided': not_provided,
            'total': total,
            'percentage': percentage
        }
        sample_stats.append({
            'sample_id': sample_id,
            'not_provided': not_provided,
            'total': total,
            'percentage': percentage
        })

# Convert to DataFrame for easier analysis
df_stats = pd.DataFrame(sample_stats)

print(f"\n=== Summary Statistics ===")
print(f"Total samples analyzed: {len(df_stats)}")
print(f"Average 'Not provided' per sample: {df_stats['not_provided'].mean():.1f}")
print(f"Average total attributes per sample: {df_stats['total'].mean():.1f}")
print(f"Average percentage 'Not provided': {df_stats['percentage'].mean():.1f}%")
print(f"Min percentage 'Not provided': {df_stats['percentage'].min():.1f}%")
print(f"Max percentage 'Not provided': {df_stats['percentage'].max():.1f}%")

Analyzing 'Not provided' values for all 17320 samples...

=== Summary Statistics ===
Total samples analyzed: 17320
Average 'Not provided' per sample: 175.7
Average total attributes per sample: 378.6
Average percentage 'Not provided': 36.6%
Min percentage 'Not provided': 0.0%
Max percentage 'Not provided': 87.3%


# Boolean attributes
Such attributes are useful for the binary classification prediction we want to build the initial model for.

In [18]:
gut_NCBI_dict.keys()

dict_keys(['ERR1073895', 'ERR1073906', 'ERR1073918', 'ERR1073929', 'ERR1073937', 'ERR1073940', 'ERR1073946', 'ERR1073952', 'ERR1073970', 'ERR1073977', 'ERR1073979', 'ERR1073993', 'ERR1079918', 'ERR1079921', 'ERR1080005', 'ERR1080025', 'ERR1080029', 'ERR1080035', 'ERR1080044', 'ERR1080049', 'ERR1080051', 'ERR1080062', 'ERR1080073', 'ERR1080087', 'ERR1080105', 'ERR1080121', 'ERR1316228', 'ERR1316229', 'ERR1316230', 'ERR1316245', 'ERR1316252', 'ERR1316253', 'ERR1316255', 'ERR1316258', 'ERR1316265', 'ERR1316266', 'ERR1316269', 'ERR1316272', 'ERR1846202', 'ERR1846205', 'ERR1846210', 'ERR1846217', 'ERR1852430', 'ERR1854662', 'ERR1854663', 'ERR1854665', 'ERR1854678', 'ERR1854681', 'ERR1854683', 'ERR1854703', 'ERR1854706', 'ERR1854710', 'ERR1854716', 'ERR1854734', 'ERR1854735', 'ERR1854746', 'ERR1854748', 'ERR1854766', 'ERR1854774', 'ERR1854778', 'ERR1854780', 'ERR1854812', 'ERR1854819', 'ERR2091828', 'ERR2091829', 'ERR2091833', 'ERR2091834', 'ERR2091835', 'ERR2091841', 'ERR2091842', 'ERR20918

In [19]:
# Get attributes for one sample
sample_attrs = gut_NCBI_dict['ERR1073895']['Attributes']

# Check data types of all attributes
print("=== Data Types of Attributes ===\n")
data_types = {}
for key, value in sample_attrs.items():
    value_type = type(value).__name__
    if value_type not in data_types:
        data_types[value_type] = []
    data_types[value_type].append(key)

for dtype, attrs in data_types.items():
    print(f"{dtype}: {len(attrs)} attributes")
    print(f"  Examples: {attrs[:3]}\n")

# Filter attributes with boolean values only ('true' or 'false')
print("\n=== Boolean-like Attributes ===\n")
boolean_attrs = {}
for key, value in sample_attrs.items():
    if isinstance(value, str) and value.lower() in ['true', 'false']:
        boolean_attrs[key] = value

print(f"Total boolean attributes: {len(boolean_attrs)}\n")
print("Boolean attributes:")
for key, value in boolean_attrs.items():
    print(f"  {key}: {value}")

# Filter attributes that are NOT boolean
print("\n=== Non-Boolean Attributes ===\n")
non_boolean_attrs = {k: v for k, v in sample_attrs.items() if not (isinstance(v, str) and v.lower() in ['true', 'false'])}
print(f"Total non-boolean attributes: {len(non_boolean_attrs)}\n")
print("Examples (first 10):")
for i, (key, value) in enumerate(list(non_boolean_attrs.items())[:10]):
    print(f"  {key}: {value}")

=== Data Types of Attributes ===

str: 482 attributes
  Examples: ['ENA first public', 'ENA last update', 'ENA-CHECKLIST']


=== Boolean-like Attributes ===

Total boolean attributes: 42

Boolean attributes:
  acne medication: false
  acne_medication_otc: false
  alcohol consumption: false
  alcohol_types_beercider: false
  alcohol_types_red_wine: false
  alcohol_types_sour_beers: false
  alcohol_types_spiritshard_alcohol: false
  alcohol_types_unspecified: true
  alcohol_types_white_wine: false
  allergic to i have no food allergies that i know of: false
  allergic_to_other: false
  allergic_to_peanuts: false
  allergic_to_shellfish: false
  allergic_to_tree_nuts: false
  allergic_to_unspecified: true
  appendix_removed: false
  cat: false
  csection: false
  dna_extracted: true
  dog: false
  has_physical_specimen: true
  lactose: false
  multivitamin: true
  nail_biter: true
  non_food_allergies_beestings: false
  non_food_allergies_drug_eg_penicillin: false
  non_food_allergies_pet

**Note:** all attributes are saved as string type.

# Subset attributes

In [20]:
# Separate attributes into two groups: those containing "subset" and those that don't
all_attributes = gut_NCBI_dict['ERR1073895']['Attributes']
subset_attributes = {k: v for k, v in all_attributes.items() if 'subset' in k.lower()}
cleaned_attributes = {k: v for k, v in all_attributes.items() if 'subset' not in k.lower()}

print(f"Total attributes: {len(all_attributes)}")
print(f"Attributes containing 'subset': {len(subset_attributes)}")
print(f"Attributes without 'subset': {len(cleaned_attributes)}\n")

# Show attributes containing "subset"
print(f"Attributes containing 'subset':")
for key, value in subset_attributes.items():
    print(f"  {key}: {value}")

Total attributes: 482
Attributes containing 'subset': 6
Attributes without 'subset': 476

Attributes containing 'subset':
  subset_age: true
  subset_antibiotic_history: true
  subset_bmi: false
  subset_diabetes: true
  subset_healthy: false
  subset_ibd: true


**Important note:**

IBD, diabetes, and antibiotic history = *TRUE* means that the participant does not have the diagnosis or did not use antibiotics in the past year.

Age = *TRUE* is for participants between 20 and 69

BMI = *TRUE* is for participants with BMI between 18.5 and 30

Healthy = *TRUE* if a participant is between 20 and 69 years old, has a BMI between 18.5 and 30, who do not report antibiotic use in more than a year, does not report IBD diagnosis, and does not report diabetes diagnosis (so all previous fields TRUE)


## Missing and unique subset values

In [21]:
# Check for NaN and unique values across ALL gut samples for subset attributes
print("=" * 100)
print("SUBSET ATTRIBUTES: Check for NaN/None values and unique values across all samples")
print("=" * 100)

subset_attrs = ['subset_age', 'subset_antibiotic_history', 'subset_bmi', 'subset_diabetes', 'subset_healthy', 'subset_ibd']

for attr_name in subset_attrs:
    values_list = []
    nan_count = 0
    
    # Collect all values for this attribute across all samples
    for sample_id in gut_NCBI_dict.keys():
        attrs = gut_NCBI_dict[sample_id]['Attributes']
        if attr_name in attrs:
            value = attrs[attr_name]
            # Check if value is NaN or None
            if value is None or (isinstance(value, float) and pd.isna(value)) or value == 'nan':
                nan_count += 1
            else:
                values_list.append(value) # if it's a valid value, add to list for unique value analysis
        else:
            nan_count += 1
    
    # Get unique values
    unique_vals = set(values_list) # use set to get unique values from the list of valid values
    
    print(f"\n{attr_name}:")
    print(f"  Total samples: {len(gut_NCBI_dict)}")
    print(f"  NaN/None values: {nan_count}")
    print(f"  Valid values: {len(values_list)}")
    print(f"  Unique values: {sorted(unique_vals)}")


SUBSET ATTRIBUTES: Check for NaN/None values and unique values across all samples

subset_age:
  Total samples: 17320
  NaN/None values: 1891
  Valid values: 15429
  Unique values: ['FALSE', 'False', 'LabControl test', 'TRUE', 'True', 'false', 'not collected', 'true']

subset_antibiotic_history:
  Total samples: 17320
  NaN/None values: 1824
  Valid values: 15496
  Unique values: ['FALSE', 'False', 'LabControl test', 'TRUE', 'True', 'false', 'not collected', 'true']

subset_bmi:
  Total samples: 17320
  NaN/None values: 1824
  Valid values: 15496
  Unique values: ['FALSE', 'False', 'LabControl test', 'TRUE', 'True', 'false', 'not collected', 'true']

subset_diabetes:
  Total samples: 17320
  NaN/None values: 1824
  Valid values: 15496
  Unique values: ['FALSE', 'False', 'LabControl test', 'TRUE', 'True', 'false', 'not collected', 'true']

subset_healthy:
  Total samples: 17320
  NaN/None values: 1824
  Valid values: 15496
  Unique values: ['FALSE', 'False', 'LabControl test', 'TRUE', '

## Proportion of True/False and missing subset attributes in the entire dataset

In [22]:
# Dictionary to store counts for each attribute
subset_counts = {}

# Count true/false values for each subset attribute across all samples
for attr_name in subset_attributes:
    true_count = 0
    false_count = 0
    missing_count = 0
    
    for sample_id in gut_NCBI_dict.keys():
        attrs = gut_NCBI_dict[sample_id]['Attributes']
        if attr_name in attrs:
            value = attrs[attr_name]
            if value.lower() == 'true':
                true_count += 1
            elif value.lower() == 'false':
                false_count += 1
            else:
                missing_count += 1
        else:
            missing_count += 1
    
    subset_counts[attr_name] = {
        'true': true_count,
        'false': false_count,
        'missing': missing_count,
        'total': true_count + false_count + missing_count
    }

# Display results
print("=" * 110)
print("SUBSET ATTRIBUTES: TRUE vs FALSE distribution across all samples")
print("=" * 110)
print(f"{'Attribute':<35} {'TRUE':<15} {'FALSE':<16} {'Missing':<17} {'Total':<10}")
print("-" * 110)

for attr_name, counts in subset_counts.items():
    true_pct = (counts['true'] / counts['total'] * 100) if counts['total'] > 0 else 0
    false_pct = (counts['false'] / counts['total'] * 100) if counts['total'] > 0 else 0
    missing_pct = (counts['missing'] / counts['total'] * 100) if counts['total'] > 0 else 0
    
    print(f"{attr_name:<35} {counts['true']:<6} ({true_pct:>5.1f}%)  {counts['false']:<6} ({false_pct:>5.1f}%)  {counts['missing']:<6} ({missing_pct:>5.1f}%)  {counts['total']:<10}")

print("=" * 110)

SUBSET ATTRIBUTES: TRUE vs FALSE distribution across all samples
Attribute                           TRUE            FALSE            Missing           Total     
--------------------------------------------------------------------------------------------------------------
subset_age                          12586  ( 72.7%)  2824   ( 16.3%)  1910   ( 11.0%)  17320     
subset_antibiotic_history           10029  ( 57.9%)  5427   ( 31.3%)  1864   ( 10.8%)  17320     
subset_bmi                          11824  ( 68.3%)  3632   ( 21.0%)  1864   ( 10.8%)  17320     
subset_diabetes                     14513  ( 83.8%)  943    (  5.4%)  1864   ( 10.8%)  17320     
subset_healthy                      6311   ( 36.4%)  9145   ( 52.8%)  1864   ( 10.8%)  17320     
subset_ibd                          13576  ( 78.4%)  1880   ( 10.9%)  1864   ( 10.8%)  17320     


# To access taxa data for a sample ID

In [29]:
# To access taxonomic data for a sample ID

gut_NCBI_dict['ERR1073895']['Taxa']

{'sk__Archaea': '0',
 'sk__Archaea;k__;p__Candidatus_Aenigmarchaeota': '0',
 'sk__Archaea;k__;p__Candidatus_Altiarchaeota;c__;o__Candidatus_Altiarchaeales': '0',
 'sk__Archaea;k__;p__Candidatus_Diapherotrites': '0',
 'sk__Archaea;k__;p__Candidatus_Micrarchaeota': '0',
 'sk__Archaea;k__;p__Candidatus_Woesearchaeota': '0',
 'sk__Archaea;k__;p__Crenarchaeota': '0',
 'sk__Archaea;k__;p__Euryarchaeota': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__;o__;f__;g__;s__Euryarchaeota_archaeon_SCGC_AAA286-E23': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Candidatus_Poseidoniia': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Halobacteria': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Halobacteria;o__Halobacteriales': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Halobacteria;o__Halobacteriales;f__Haloarculaceae': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Halobacteria;o__Halobacteriales;f__Haloarculaceae;g__Halomicroarcula': '0',
 'sk__Archaea;k__;p__Euryarchaeota;c__Halobacteria;o__Halobacteriales;f__Haloarcu

sk__: superkingdom

k__ : kingdom

p__ : phylum

c__ : class

o__ : order

f__ : family

g__ : genus

s__ : species


Taxa contains the full taxonomic tree paths, while Phylum is an aggregated summary at the phylum level. Ideally the children nodes should sum up to the parent node but that's not always the case due to measurement errors or not detecting at every level or not detecting every possible species.

Your dataset is essentially a flattened taxonomy tree. Each feature corresponds to one node in that tree.

Feature 1: p__Firmicutes

Feature 2: p__Firmicutes;c__Clostridia

Feature 3: p__Firmicutes;c__Clostridia;o__Clostridiales

Feature 4: p__Firmicutes;c__Clostridia;o__Clostridiales;g__Faecalibacterium

# To access phylum data for a sample ID

In [30]:
gut_NCBI_dict['ERR1073895']['Phylum']

{'Candidatus_Aenigmarchaeota': '0',
 'Candidatus_Altiarchaeota': '0',
 'Candidatus_Diapherotrites': '0',
 'Candidatus_Micrarchaeota': '0',
 'Candidatus_Woesearchaeota': '0',
 'Crenarchaeota': '0',
 'Euryarchaeota': '0',
 'Thaumarchaeota': '0',
 'Unassigned': '0',
 'Acidobacteria': '0',
 'Actinobacteria': '501',
 'Aquificae': '0',
 'Armatimonadetes': '0',
 'Bacteroidetes': '5315',
 'Balneolaeota': '0',
 'Caldiserica': '0',
 'Calditrichaeota': '0',
 'Candidatus_Abawacabacteria': '0',
 'Candidatus_Adlerbacteria': '0',
 'Candidatus_Amesbacteria': '0',
 'Candidatus_Atribacteria': '0',
 'Candidatus_Azambacteria': '0',
 'Candidatus_Beckwithbacteria': '0',
 'Candidatus_Berkelbacteria': '0',
 'Candidatus_Bipolaricaulota': '0',
 'Candidatus_Buchananbacteria': '0',
 'Candidatus_Campbellbacteria': '0',
 'Candidatus_Chisholmbacteria': '0',
 'Candidatus_Cloacimonetes': '0',
 'Candidatus_Collierbacteria': '0',
 'Candidatus_Colwellbacteria': '0',
 'Candidatus_Curtissbacteria': '0',
 'Candidatus_Dadaba

# Saving the gut dictionary

In [ ]:
# Save gut_NCBI_dict
with open('data/raw/gut_NCBI_dict', 'wb') as file:
    pickle.dump(gut_NCBI_dict, file)

print(f"✓ Saved gut_NCBI_dict ({len(gut_NCBI_dict):,} samples) to 'data/raw/gut_NCBI_dict'")

✓ Saved gut_NCBI_dict (17,320 samples) to 'gut_NCBI_dict'
